### ISSUE TITLE
MLX Migration Results: 120x faster, 12x less memory

### ISSUE THREAD

**Issue by pierretokns** [pierretokns/agenthub#14]
### Benchmark Results (M2 Ultra, 64GB)

| Metric | PyTorch MPS | MLX Native | Improvement |
|--------|------------|------------|-------------|
| Step time (batch=128, seq=256) | ~31,000ms | **251ms** | **~120x** |
| Step time (batch=64, seq=256) | ~15,000ms | **194ms** | **~77x** |
| Peak memory (training) | 79 GB | **6.4 GB** | **12x less** |
| Peak memory (inference) | ~2 GB | **0.57 GB** | **3.5x less** |
| JIT compile (first step) | N/A | 1032ms | One-time cost |
| MTEB eval (STSBenchmark) | ~60s | ~30s | **2x** |

### What Changed
- **`src/model.py`**: Full ModernBERT encoder (22 layers) implemented natively in MLX
  - Fused QKV projections, GLU MLP, alternating local/global attention
  - RoPE with different theta for local vs global layers
  - `mx.fast.scaled_dot_product_attention()` for native Metal attention
  - Loads weights directly from HF safetensors (no PyTorch conversion needed)
- **`src/train.py`**: Complete rewrite from PyTorch to MLX
  - `nn.value_and_grad()` replaces `loss.backward()` + `optimizer.step()`
  - `mlx.optimizers.AdamW` replaces `torch.optim.AdamW`
  - Removed all PyTorch boilerplate: `.to(device)`, `autocast`, `empty_cache`, `GradScaler`
  - Hard negative mining uses numpy (from `encode_sentences()`)
- **`pyproject.toml`**: Removed `sentence-transformers` dependency (pulls in PyTorch)

### Assets Uploaded
- [pierretokns/ModernBERT-base-mlx](https://huggingface.co/pierretokns/ModernBERT-base-mlx) — MLX-compatible weights on HF
- [pierretokns/mteb-contrastive-clean-v1](https://huggingface.co/datasets/pierretokns/mteb-contrastive-clean-v1) — 89K decontaminated training pairs

### What This Means for Research Velocity
With the old PyTorch MPS pipeline:
- 4-stage experiment: ~51 min training + 10 min eval = **~61 min/experiment**
- Bottleneck: contrastive stage at 31s/step, most time spent on CPU↔GPU sync

With MLX:
- At 251ms/step, the same ~3500 steps take **~15 min** instead of 30 min
- Hard neg mining much faster (numpy matmul, no GPU transfer)  
- Estimated: **~25-30 min/experiment** (2x more experiments per hour)
- Memory headroom: 6.4GB peak vs 64GB available = room for batch_size=512+

### Community Resources Used
- [mlx-community ModernBERT collection](https://huggingface.co/collections/mlx-community/modernbert-67eea1cbffcd7f6fd1c09217)
- [pappitti/modernbert-mlx](https://github.com/pappitti/modernbert-mlx) — reference implementation
- [Blaizzy/mlx-embeddings](https://github.com/Blaizzy/mlx-embeddings) — MLX embedding models
- [jina-ai/mlx-retrieval](https://github.com/jina-ai/mlx-retrieval) — contrastive training on MLX
- `autoresearch-mlx` — our own repo for `value_and_grad` patterns, custom AdamW

### Current Best Score (pre-MLX)
```
primary_score: 27.20 (exp-3, only clean/kept result)
```
Next step: run first MLX experiment to establish new baseline.

Closes #13
